### PR TITLE
Hashing and Cryptography

### LINKED ISSUE #1331: Add section on hashing to Book
The book currently doesn't highlight hashing, what's available for hashing etc.

I think that would be a nice section to include. I'll try to keep examples light, as once we have better generic support, the syntax will change.

### PR DESCRIPTION
**Pull request by SilentCicero**
Closes #1331
- Adds a simple hashing and cryptography section
- Shows how to get public keys and addresses

### CODE REVIEW

**Comment by SilentCicero** [comment]
Note, in order to make the `forc-documentor` pass I had to include these three forc file changes for CI. Mitch said it was okay. Don't hurt me.

**Review by mitchmindtree — APPROVED** [review]

**Comment by mitchmindtree** [comment]
cc @binggh curiously our `cargo run --bin forc-documenter write-docs --dry-run` CI job appears to be giving an error during CI, but I'm unable to recreate the error locally and running `cargo run --bin forc-documenter write-docs` doesn't result in any changes :thinking: any ideas what might be happening here?

**Comment by mitchmindtree** [comment]
It looks like this same error is occurring `master` too.

I think what's potentially happening is that there might be a bug where the `forc-documenter` doesn't know how to deal with the new `- Dependencies` entry in the `summary.md` and is failing.

That said, I wonder why that wasn't caught by CI in https://github.com/FuelLabs/sway/pull/1340.

Investigating now.

**Comment by mitchmindtree** [comment]
> bug where the forc-documenter doesn't know how to deal with the new - Dependencies entry in the summary.md

I take it back - it looks like the `forc --help` output is now coming from `stderr` rather than `stdout` for some reason :thinking:

**Comment by binggh** [comment]
Hmm, it's weird it's not printing the `forc` version that is running. The script should print something like this:

`Running forc --help using forc 0.10.3`

before running by calling `forc --version`, you can see this behaviour in #1340 as well.

**Comment by mitchmindtree** [comment]
Yeah it looks like all of the output from `forc` is suddenly going to `stderr` instead of `stdout`! Not sure what's going on here yet :woozy_face:

**Review by mitchmindtree — APPROVED** [review]
OK should be fine to land this now whenever you're ready @SilentCicero :)

File: docs/src/blockchain-development/hashing_and_cryptography.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Calling Contracts
+
+The Sway standard library provides easy access to hashing (namely `sha256` and Ethereum compatible `keccak256`) operations and Ethereum-compatible Secp256k1-based signature recovery operations.
```
Comment (adlerjohn) [inline docs/src/blockchain-development/hashing_and_cryptography.md:3]:
```suggestion
The Sway standard library provides easy access to a selection of cryptographic hash functions (`sha256` and Ethereum-compatible `keccak256`), and Ethereum-compatible secp256k1-based signature recovery operations.
```

File: docs/src/blockchain-development/hashing_and_cryptography.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# Calling Contracts
```
Comment (adlerjohn) [inline docs/src/blockchain-development/hashing_and_cryptography.md:1]: Title

File: examples/hashing/Forc.lock [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+name = 'core'
+dependencies = []
+
+[[package]]
+name = 'hashing'
+dependencies = ['std git+https://github.com/fuellabs/sway?tag=v0.10.3#e7676db6f4ebc7a3885f87514d16a703a99410d7']
+
+[[package]]
+name = 'std'
+source = 'git+https://github.com/fuellabs/sway?tag=v0.10.3#e7676db6f4ebc7a3885f87514d16a703a99410d7'
```
Comment (adlerjohn) [inline examples/hashing/Forc.lock:11]: Regenerate this file

File: examples/hashing/src/main.sw [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    // Hash a single b256 value.
+    let hashed_b256 = hash_value(hashed_u64, HashMethod::Keccak256);
+
+    // Hash two b256 values.
+    let hashed_pair = hash_pair(hashed_b256, VALUE_A, HashMethod::Sha256);
+
+    log_b256(hashed_pair);
+}
+
```
Comment (adlerjohn) [inline examples/hashing/src/main.sw:25]: Format this file.

File: examples/signatures/src/main.sw [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    // A recovered Fuel address.
+    let result_address: Result<Address, EcRecoverError> = ec_recover_address(signature, MSG_HASH);
+    if let Result::Ok(address) = result_address {
+        log_b256(address.value);
+    } else {
+        panic(0);
+    };
+}
+
```
Comment (adlerjohn) [inline examples/signatures/src/main.sw:32]: Format this file. Please turn on format on save in your IDE.

**Review by adlerjohn — CHANGES_REQUESTED** [review]

**Comment by mitchmindtree** [comment]
@SilentCicero just a heads up - as this branch was force-pushed from my end when applying the rebase and addressing the CI issues, you should delete your existing local branch and re-fetch this branch before addressing John's comments.

File: docs/src/blockchain-development/hashing_and_cryptography.md (hunk 2 above)
Comment (SilentCicero) [reply docs/src/blockchain-development/hashing_and_cryptography.md:1]: I was getting pretty tired with this one... Damn.

File: examples/signatures/src/main.sw (hunk 5 above)
Comment (SilentCicero) [reply examples/signatures/src/main.sw:32]: Do we have instructions for that somewhere? Can't seem to find any for the Sway formatter particularly.

Comment (adlerjohn) [reply examples/signatures/src/main.sw:32]: If you're using vscode and have format in save turned on in general, then it should just work automatically.

Comment (SilentCicero) [reply examples/signatures/src/main.sw:32]: Okay, I think I'd like to notate that in the docs. As it's not immediately clear its available by reading the docs that the plugin is capable of this. But it's really cool!

Comment (SilentCicero) [reply examples/signatures/src/main.sw:32]: Also, my only hesitation with doing this as well, is that currently the formatting for dependancies / use statements is not pleasant. Until that's resolved, I'll manually run the formatter. But I'll be sure to do though every time.

Comment (adlerjohn) [reply examples/signatures/src/main.sw:32]: It doesn't really matter that it's not pleasant. There are no guarantees that you or anyone else will _personally_ always find the formatted code pleasant irrespective of this specific case. Please turn on format on save.

Comment (SilentCicero) [reply examples/signatures/src/main.sw:29]: Until the dependency formatting is resolved, I'll be running it manually.

Formatting all my dependancies in one giant line is not only not pleasant, it's unsafe and should be resolved first.

Comment (adlerjohn) [reply examples/signatures/src/main.sw:32]: Not exactly sure what's unsafe about that.

Until #1209 is resolved, a workaround would be to split it up into multiple `use` statements.

**Review by adlerjohn — COMMENTED** [review]
Update from `master` to pick up changes from #1384

**Comment by mohammadfawaz** [comment]
I'm going to push some changes into this branch shortly.

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**